### DISCUSSION TITLE
Something is calling ConvertBack with null

### DISCUSSION THREAD

**Discussion by Mrcubix** [AvaloniaUI/Avalonia#9115] [Q&A] [answered]
The example here is used in an `Avalonia.xplat` template, which is powered using 11.0.0-Preview1

```cs
public class WordedNumberToIndex : IValueConverter
{
     public Dictionary<uint, string> SomeDictionary { get; set; } = new() 
     {
        [1] = "One",
        [2] = "Two"
     };

    public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        if (value is string wordedNumber)
        {
            return SomeDictionary.Values.IndexOf(wordedNumber);
        }

        return -1;
    }

    public object? ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        if (value is int index)
        {
            if (SomeDictionary.Count > index && index >= 0)
                    return SomeDictionary.Values.ToArray()[index];
        }

        return -1;
    } 
}
```
## What i'm trying to do

I'm trying to set a the selected value of a `ComboBox` to one, saved elsewhere, in a file for example, here illustrated by `DefaultValue`

```cs
public class ExampleObject
{
    public uint Number { get; set; }
}

public class SomeViewModel : ViewModelBase
{
    public Dictionary<uint, string> SomeDictionary { get; set; } = new() 
     {
        [1] = "One",
        [2] = "Two"
     };
    
    public string _defaultValue = "One";
    public string DefaultValue 
    {
        get => _defaultValue;
        set => this.RaiseAndSetIfChanged(ref _defaultValue, value);
    }
}
```

```xml
<ComboBox Items="{Binding SomeDictionary}" SelectedIndex="{Binding DefaultValue, Converter={StaticResource WordedNumberToIndex}}" Width="320">
  <ComboBox.ItemTemplate>
    <DataTemplate>
      <ComboBoxItem>
        <ComboBoxItem.Content>
          <StackPanel Orientation="Horizontal">
            <TextBlock Text="{Binding Number}" />
            <TextBlock Text="-" />
            <TextBlock Text="{Binding Number, Converter={StaticResource NumberToText}}" />
          </StackPanel>
        </ComboBoxItem.Content>
      </ComboBoxItem>
    </DataTemplate>
  </ComboBox.ItemTemplate>
</ComboBox>
 ```

## The issue

I'm currently having an issue with something, that isn't part of my code, calling `ConvertBack` with `-1` as argument.

just for clarification, `DefaultValue` is never `null`, well that is, it's never `null` until something that isn't part of my code call `ConvertBack` then proceed to set the value of `DefaultValue` with `null`, causing the ComboBox to show the following:

` - `

instead of, for example:

`1 - one` (when `DefaultValue` is `"One"`)

at some point in my code, i call `this.RaiseAndSetIfChanged(ref _defaultValue, value);` from an external event, `value` **is not** `null`, and has a value of `"One"`

`DefaultValue`'s Setter is then called, but after that, something is calling `ConvertBack` with -1 as argument, which is weird.

any ideas?

# An Update

I figured out that if i replace `someDictionary`'s reference, even if i don't `null` the variable myself, something will temporarily figure out that `SomeDictionary` is `null` and try to call `ConvertBack` w/ `null`, once the reference is replaced, something will not attempt to update the ComboBox

this effectively seems like a bug to me, this need confirmation

```cs
public void SomeTask
{
    // this cause something to temporarily think SomeDictionary is null, and doesn't try to update it later on
     SomeDictionary = new() 
     {
        [1] = "One",
        [2] = "Two"
     };
}

**Comment by timunie** [comment, marked as answer]
Can you upload this as a small sample to a bew GitHub repo?

**Reply by Mrcubix** [reply]
Sure, here you go, made sure the issue is present https://github.com/Mrcubix/Reference-Issue

**Reply by timunie** [reply]
Okay I think I found the issue. You are replacing the items. In that case the ComboBox will also loose the selection. If you tell the UI that also the selection changed it works as expected I think. Please test the PR I sent you. 

```diff
public void ChangeReference()
{
    SomeDictionary = new() 
    {
        [1] = "One",
        [2] = "Two"
    };
    
+    this.RaisePropertyChanged(nameof(DefaultValue));
}
```

**Reply by Mrcubix** [reply]
as i've said in the PR, this isn't really ideal but this will do as a possible fix

**Reply by timunie** [reply]
I agree with you, and I think this issue https://github.com/AvaloniaUI/Avalonia/issues/9046 is related. Again, if you want to help us, clone Avalonia and try to provide a fix. Your help is welcome as far as I can tell.

**Reply by Mrcubix** [reply]
The hard part is doing that as the 502 bad gateway issue on Avalonia's nuget feed is probably still a thing
